engine/src/org/pentaho/di/trans/steps/xmlinputstream/XMLInputStream.java

```diff
@@ -142,16 +108,86 @@ public boolean processRow( StepMetaInterface smi, StepDataInterface sdi ) throws
       setOutputDone();
       return false;
     }
+    return true;
+  }
 
+  private boolean openNextFile() throws KettleException {
+    try {
+      closeFile();
+      if ( data.filenr >= data.filenames.length ) {
+        return false;
+      }
+      data.fileObject = KettleVFS.getFileObject( data.filenames[data.filenr], getTransMeta() );
+      data.inputStream = KettleVFS.getInputStream( data.fileObject );
+      data.xmlEventReader = data.staxInstance.createXMLEventReader( data.inputStream, data.encoding );
+    } catch ( IOException e ) {
+      throw new KettleException( e );
+    } catch ( XMLStreamException e ) {
+      throw new KettleException( e );
+    }
+    data.filenr++;
+    if ( meta.isAddResultFile() ) {
+      // Add this to the result file names...
+      ResultFile resultFile =
+        new ResultFile( ResultFile.FILE_TYPE_GENERAL, data.fileObject, getTransMeta().getName(), getStepname() );
+      resultFile.setComment( BaseMessages.getString( PKG, "XMLInputStream.Log.ResultFileWasRead" ) );
+      addResultFile( resultFile );
+    }
     return true;
   }
 
+  private void closeFile() throws KettleException {
+    try {
+      if ( data.xmlEventReader != null ) {
```

Comment:
1- I would like to see a catch around each of these
2- Usually, when we get an exception closing files, we simply log it as a warning and move on. There's nothing more to do if you cannot close the file or input stream.